core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -577,14 +582,42 @@ private SnapshotInfo inProgressSnapshot(SnapshotsInProgress.Entry entry) {
                         shardSnapshotStatus.updateStage(IndexShardSnapshotStatus.Stage.FAILURE);
                         shardSnapshotStatus.failure(shardFailure.reason());
                         shardStatus.put(shardId, shardSnapshotStatus);
+                        hasFailedShards = true;
                     } else {
-                        IndexShardSnapshotStatus shardSnapshotStatus =
-                            repository.getShardSnapshotStatus(snapshotInfo.snapshotId(), snapshotInfo.version(), indexId, shardId);
-                        shardStatus.put(shardId, shardSnapshotStatus);
```

Comment:
@imotov is correct, we only need to check the snapshot status here and execute the read based on it, I pushed https://github.com/elastic/elasticsearch/pull/23833/commits/45c994d98b955c834aa8533de21e54d3798199c5